pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentReplicator.java

```diff
@@ -199,11 +199,11 @@ protected void disableReplicatorRead() {
      * Calculate available permits for read entries.
      *
      * @return
-     *   0:  Producer queue is full, no permits.
-     *  -1:  Rate Limiter reaches limit.
-     *  >0:  available permits for read entries.
+     *   0, O:  Producer queue is full, no permits.
+     *  -1, -1:  Rate Limiter reaches limit.
+     *  >0, >0:  available permits for read entries.
      */
-    private int getAvailablePermits() {
+    public Pair<Integer, Long> getAvailablePermits() {
```

Comment:
It's better to use a record than to use Pair. Pair results in hard-to-read code. An additional reason here is that Integer and Long are boxed objects which are adding additional overhead compared to a record.

Reply:
I can add a class to hold the available messages and bytes, so like:

```
    @Data
    @AllArgsConstructor
    private static class AvailablePermits {
        private int messages;
        private long bytes;
    }
```

`record` requires the JDK 14, we cannot cherry-pick this PR to the previous version.

Reply:
Ping @lhotari